### PR TITLE
Feature solve watershed

### LINKED ISSUE #23: [feature] BMP pollutant load IO
must support structural facilities and site-based treatment

build load reduction functions from the reference data #22 
map project BMP types to the function names
use project configuration file to define pollutants under analysis OR infer from input file to run all pollutants from reference data

### LINKED ISSUE #24: [feature] load reduction algorithm
already drafted and reviewed - not merged

needs to be improved to:
 - compute treatment performance from flow-based bmps (nomograph improvement needed)
 - handle online-offline toggle for flow-based and volume based bmps. 
 - handle dry weather flow considerations (maybe monthly, apr - sept)
 - handle baseline (this _should_ be handled by sending only a network of RSBs into nereid which don't contain BMPs, then the difference from baseline is computed per RSB.) need to check this with benevolent overlords.

### LINKED ISSUE #25: [feature] BMP volume capture performance
Link BMP to appropriate nomograph(s) via configuration files. BMP json has a ref_data_key == lspc basin code. use lspc basin id to link a reference data nomograph.

in network solution route, initialize necessary nomographs at the top of the network solution algorithm and proceed.

 - handle flow-based nomographs (treatment only)
 - handle volume based nomographs (treatment, treatment + retention, retention). this code is 90% complete
 - handle dry weather flows
 - handle online/offline

### PR DESCRIPTION
**Pull request by austinorr**
- Implements volume based and flow based volume solutions closes #25 and #26 
- implements load reduction algorithms for wet and dry weather closes #24 
- implements influent and effluent curves informed by the BMP database closes #23

### CODE REVIEW

**Comment by phobson** [comment]
@austinorr I can take a look later today

**Comment by austinorr** [comment]
ok cool, don't invest too much yet. I'm working on a pr with a few more tests and more documentation + incorporates the feedback from your last review. I need to leave more breadcrumbs through the graph traversal, since there are several routes depending on what the node type is.

**Comment by austinorr** [comment]
couple quick notes:
the diff looks huge (10k+) but over half of that is updated reference files, like the land_surface_data file. these example reference files are truncated considerably compared to their real versions, but they are still quite large files.

I have sequestered the entire json file diff into a single commit so it's easier to skip.

**Comment by austinorr** [comment]
@phobson i think this is ready to review now. docs are improved throughout, especially in `src.watershed`

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,72 @@
+MAKEFLAGS += --silent
```
[the remark this answers is not in the document]
Comment (austinorr) [reply Makefile:1]: you can run it on your mac. i made it so it was easier for me to develop this at home on my own machine.

File: nereid/nereid/api/api_v1/endpoints/treatment_facilities.py [hunk 2]
```diff
@@ -123,7 +121,9 @@ async def initialize_treatment_facility_parameters(
     treatment_facilities, context = tmnt_facility_req
 
     task = bg.background_initialize_treatment_facilities.s(
-        treatment_facilities=treatment_facilities.dict(), context=context
+        treatment_facilities=treatment_facilities.dict(),
+        pre_validated=True,
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/api/api_v1/endpoints/treatment_facilities.py:125]: checkout the function args to the `initialize_treatment_facility_parameters` function. in fastapi we can inject functions that catch the data that gets posted to our routes by using the `fastapi.Depends` function. in this case, i'm injecting the result of a validator function defined right before this function, called `validate_facility_request`. This intercepts all the args that are posted to the endpoint (the treatment_facility object and the context, which also has a separate dependency) and runs the `validate_treatment_facility_models` script on the unvalidated data before passing it to the endpoint.

all of this is to ensure that before we pass bad data to the background process (where exceptions register to the client as 'internal server errors' and are utterly useless) we perform validation on the submission to identify errors and return a json object detailing the error, rather than just a 500 error code.

File: nereid/nereid/data/default_data/state/region/config.yml [hunk 3]
```diff
@@ -90,18 +108,91 @@ api_recognize:
...
+    pollutant_column: pollutant
+    preprocess:
+      - remaps:
+        - left: unit
+          right: -- # with how = replace, the right key is not parsed at all.
+          how: replace  # interpreted as 'left' = mapping['left']
+          mapping:
+            # units must be compatible with the `pint` library
+            "MPN/100mL": "MPN/_100ml"
+            "MPN/100 mL": "MPN/_100ml"
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/data/default_data/state/region/config.yml:194]: yes, case matters, but it's addressed in line 4 of the config.yml file which remaps MPN = count = mpn and _100ml = 100 * milliliter = _100mL. 

these `pint_unit_registry` mappings are loaded by the `update_unit_registry` decorator function that wraps any task that contains code that does a unit conversion.

also, we do not need to be more explicit or tolerant of case on lines 193-194 or 207-208 because the project config should actually know what the terms are in the project data files, that's why it exists.

File: nereid/nereid/src/land_surface/loading.py [hunk 4]
```diff
@@ -32,76 +37,154 @@ def detailed_volume_loading_results(df: pandas.DataFrame,) -> pandas.DataFrame:
     df["imp_eff_area_acres"] = df["imp_ro_coeff"] * df["imp_area_acres"]
     df["perv_eff_area_acres"] = df["perv_ro_coeff"] * df["perv_area_acres"]
     df["eff_area_acres"] = df["imp_eff_area_acres"] + df["perv_eff_area_acres"]
+    df["ro_coeff"] = df["eff_area_acres"] / df["area_acres"]
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/land_surface/loading.py:40]: yes, i think that your definition is correct. we usually deal with it in the form of `effective_area = ro_coeff * total_project_area` which is what you see on lines 37 and 38 for the imp area and the perv area. our LSPC modeling effort gives us an impervious surface ro coeff, and a pervious surface ro coeff for each type of landuse. This portion of the volume loading calculation helps to combine those two model results into one value.

File: nereid/nereid/src/nomograph/interpolators.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
+
+import matplotlib.pyplot as plt
+import numpy
+import pandas
+from matplotlib.axes import Axes
+from scipy.interpolate import CloughTocher2DInterpolator as CT2DI
+
+
+def bisection_search(
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/nomograph/interpolators.py:10]: the kicker is that we have to transform f so that we ensure a and b have different signs before we could use `optimize.bisect`, but in this case our function is a slice of a 2dcloughtocher interpolation, and i don't know how to transform it such that the value i 'know' is =0 for the value i want back.

if this can be made to work with a scipy built in, i'd actually prefer it, but for now it converges in 4-8 iterations, so it's pretty fast for this use case.

File: nereid/nereid/src/treatment_site/tasks.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        )
+    }
+
+    tmnt_sites = []
+
+    if sites:
+
+        _df = pandas.DataFrame(sites)
+
+        remaider_data = []
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/treatment_site/tasks.py:34]: yes, i was trying to type that.

File: nereid/nereid/core/utils.py [hunk 7]
```diff
@@ -86,3 +89,66 @@ def get_request_context(
     request_context["data_path"] = str(data_path)
 
     return request_context
+
+
+def validate_models_with_discriminator(
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/core/utils.py:94]: [discriminator](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#fixed-fields-20) is a field used to indicate to the validator which object it should become. thus, provided a list of objects with a string field that indicates what they are, and a mapping from strings to object models, i can connect a list of unvalidated data objects to the correct validator even if each item in the list should have a different validation function.

for whatever reason, this has not been fully baked into `fastapi` or `pydantic` which is used internally by `fastapi` for validation. see the discussion that inspired my implementation [here](https://github.com/samuelcolvin/pydantic/issues/619)

File: nereid/nereid/tests/test_src/test_watershed/test_solve_watershed_sequence.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+from nereid.src.land_surface.tasks import land_surface_loading
+from nereid.src.network.tasks import solution_sequence
+from nereid.src.network.utils import nxGraph_to_dict, graph_factory
+from nereid.src.watershed.tasks import solve_watershed
+from nereid.tests.utils import check_results_dataframes, attrs_to_resubmit
+
+
+@pytest.mark.parametrize("pct_tmnt", [0.0, 0.6])
+@pytest.mark.parametrize("n_nodes", [100])
+def test_watershed_solve_sequence(contexts, watershed_requests, n_nodes, pct_tmnt):
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/tests/test_src/test_watershed/test_solve_watershed_sequence.py:16]: ok, call me ; )

File: nereid/nereid/src/land_surface/loading.py [hunk 9]
```diff
@@ -32,76 +37,154 @@ def detailed_volume_loading_results(df: pandas.DataFrame,) -> pandas.DataFrame:
...
+) -> pandas.DataFrame:
+    """This function aggregates the dry weather flowrate by season according
+    to the config file spec.
+    """
+
+    developed_area_acres = df.get("developed_area_acres", 0.0)
+
+    for season, months in seasons.items():
+        if months is not None:
+            dwf_cfs_ac = df.get(f"{season}_dry_weather_flow_cuft_psecond_pacre", 0.0)
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/land_surface/loading.py:59]: well, `dwf_cf_ps_pac` would be the *most* consistent, lol.

i don't care which, `pac` or `per_ac`. it's your comment, you choose your favorite correction. : )

File: nereid/nereid/src/land_surface/loading.py [hunk 10]
```diff
@@ -3,13 +3,18 @@
...
 
-from nereid.core.units import ureg
+from nereid.core.utils import safe_divide
+from nereid.core.units import Constants
 
 # TODO: implement fallback to e.g., "TRANS-B-10" rather than "401070-TRANS-B-10"
 # in case ref data is missing.
 
 
 def clean_land_surface_dataframe(df: pandas.DataFrame) -> pandas.DataFrame:
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/land_surface/loading.py:13]: did you mean this function: `detailed_loading_results`? the one you tagged expects any df with an 'imp_pct' column and an 'area_acres' column, so the answer there would be a bit boring.

the better story is that the function you tagged is part 1 in the detailed loading results pattern, which requires a df of your land surfaces request object, which are a list of these:
```
class LandSurface(BaseModel):
    node_id: str
    surface_key: str = Field(..., example="104506-RESSFH-B-5")
    area_acres: float
    imp_area_acres: float
```
joined to your project reference data, which are these (for the default data land_surface_data.json):
```
  "data":[
    {
      "surface_id":"10101000-COMM-A-0",
      "perv_ro_depth_inches":2.3145148876,
      "imp_ro_depth_inches":11.2305913657,
      "perv_ro_coeff":0.1915657302,
      "imp_ro_coeff":0.929523697
    },
```

plus whatever the creator of the config file joined in the `preprocess` sections....

in short, i've separated these three functions into pieces so that the structure of the data they handle is evident by the <20 line bodies of each. each function requires a dataframe with the keys accessed inside the function, and if it fails it's *most likely* because of a bad config file that should be debugged on project initialization.

at any rate, how would we show an example df in the docstring? are the required fields not enough to indicate the structure of the flat table that we are passing around?

File: nereid/nereid/src/land_surface/loading.py [hunk 11]
```diff
@@ -32,76 +37,154 @@ def detailed_volume_loading_results(df: pandas.DataFrame,) -> pandas.DataFrame:
...
-        poc = param["short_name"]
-        emc_col = "_".join([poc, "conc", conc_unit.lower().replace("_", "")])
-        load_col = "_".join([poc, "load", load_unit.lower().replace("_", "")])
+    It also contains derived values from 'src.wq_parameters.init_wq_paramters'
+    """
 
-        if emc_col in df:
-            factor = (ureg("cubic_feet") * ureg(conc_unit)).to(load_unit).magnitude
-            df[load_col] = df["runoff_volume_cuft"] * df[emc_col] * factor
+    for param in wet_weather_parameters:
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/land_surface/loading.py:94]: wet weather means rain is falling, and the hydrology of facility capture conditions can be related to a nomograph of long-term capture efficiency.

land surface washoff event mean concentrations, or EMC's, also implicitly imply that they are pertinent during rain events. long term EMC's *can* differentiate between summer and winter since the pollutant build up time is different, but they often don't -- or if they do, there's not a good enough data sample size for robust statistics.

Dry weather flows (say from irrigation, or from groundwater interflow) are not coupled to rainfall events, and it might be different in winter time than they are in summer time. dry weather is defined as '3 days after rainfall' and each month of the year has a different average number of dry days defined in the 'dry_weather_flow.json' file. by multiplying the number of dry days by the dry weather flow rate for the season & basin we get dry weather volume for that month.

File: nereid/nereid/src/land_surface/loading.py [hunk 12]
```diff
@@ -114,15 +197,22 @@ def summary_loading_results(
...
 
-    for param in parameters:
-        conc_unit = param["concentration_unit"]
-        load_unit = param["load_unit"]
-
-        poc = param["short_name"]
-        load_col = "_".join([poc, "load", load_unit.lower().replace("_", "")])
-        conc_col = "_".join([poc, "conc", conc_unit.lower().replace("_", "")])
-        factor = (ureg(load_unit) / ureg("cubic_feet")).to(conc_unit).magnitude
+    for param in wet_weather_parameters:
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/land_surface/loading.py:200]: I'm confused about this comment. what are we discussing here?

File: nereid/nereid/src/land_surface/tasks.py [hunk 13]
```diff
@@ -14,28 +15,49 @@
 def land_surface_loading(
     land_surfaces: Dict[str, Any], details: bool, context: Dict[str, Any]
 ) -> Dict[str, List]:
+    """computes loading for volume runoff and pollutants for each land
+    surface 'sliver' and aggregates values for each node. Returning results
+    for the slivers is toggled by the `details` kwarg. if 'true' the resonse
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/land_surface/tasks.py:20]: mind your ps and qs...

File: nereid/nereid/src/network/utils.py [hunk 14]
```diff
@@ -93,6 +93,36 @@ def thin_graph_dict(graph_dict: Dict[str, Any]) -> Dict[str, Any]:
 
 
 def nxGraph_to_dict(g: nx.Graph) -> Dict[str, Any]:
+    """Convert a networkx garph object into a dictionary
```
[the remark this answers is not in the document]
Comment (austinorr) [reply nereid/nereid/src/network/utils.py:96]: maybe someday. i'm generally not a fan of doctests... nothing against them, just hate fiddling with code that lives its whole life as a comment.

plus this function is tested to what i consider an 'overkill' degree since things like 'multi graph' and 'undirected graph' support are very much considered 'invalid watersheds' for our purposes.

[the remark this answers is not in the document]
Comment (phobson) [reply nereid/nereid/src/network/utils.py:96]: That's fair. What I do like about doctests is that they ensure your docs are correct. "wrongs docs are worse than no docs" and whatnot
